staticaddr/loopin/fsm.go

```diff
@@ -0,0 +1,355 @@
+	OnRecover                       = fsm.EventType("OnRecover")
+)
+
+// LoopInStatesV0 returns the state and transition map for the loop-in state
+// machine.
+func (f *FSM) LoopInStatesV0() fsm.States {
+	return fsm.States{
+		fsm.EmptyState: fsm.State{
+			Transitions: fsm.Transitions{
+				OnInitHtlc: InitHtlc,
+			},
+			Action: fsm.NoOpAction,
+		},
+		InitHtlc: fsm.State{
+			Transitions: fsm.Transitions{
+				OnHtlcInitiated: SignHtlcTx,
+				OnRecover:       UnlockDeposits,
+				fsm.OnError:     UnlockDeposits,
+			},
+			Action: f.InitHtlcAction,
+		},
+		SignHtlcTx: fsm.State{
+			Transitions: fsm.Transitions{
+				OnHtlcTxSigned: MonitorInvoiceAndHtlcTx,
+				OnRecover:      UnlockDeposits,
+				fsm.OnError:    UnlockDeposits,
+			},
+			Action: f.SignHtlcTxAction,
+		},
+		MonitorInvoiceAndHtlcTx: fsm.State{
+			Transitions: fsm.Transitions{
+				OnPaymentReceived:  PaymentReceived,
+				OnSweepHtlcTimeout: SweepHtlcTimeout,
+				OnSwapTimedOut:     UnlockDeposits,
+				OnRecover:          MonitorInvoiceAndHtlcTx,
+				fsm.OnError:        UnlockDeposits,
+			},
+			Action: f.MonitorInvoiceAndHtlcTxAction,
+		},
+		SweepHtlcTimeout: fsm.State{
+			Transitions: fsm.Transitions{
+				OnHtlcTimeoutSweepPublished: MonitorHtlcTimeoutSweep,
+				OnRecover:                   SweepHtlcTimeout,
+				fsm.OnError:                 Failed,
+			},
+			Action: f.SweepHtlcTimeoutAction,
+		},
+		MonitorHtlcTimeoutSweep: fsm.State{
+			Transitions: fsm.Transitions{
+				OnHtlcTimeoutSwept: HtlcTimeoutSwept,
+				OnRecover:          MonitorHtlcTimeoutSweep,
+				fsm.OnError:        Failed,
+			},
+			Action: f.MonitorHtlcTimeoutSweepAction,
+		},
+		PaymentReceived: fsm.State{
+			Transitions: fsm.Transitions{
+				OnFetchSignPushSweeplessSweepTx: FetchSignPushSweeplessSweepTx,
+				OnRecover:                       SucceededSweeplessSigFailed,
+				fsm.OnError:                     SucceededSweeplessSigFailed,
+			},
+			Action: f.PaymentReceivedAction,
+		},
+		FetchSignPushSweeplessSweepTx: fsm.State{
+			Transitions: fsm.Transitions{
+				OnSweeplessSweepSigned: Succeeded,
+				OnRecover:              SucceededSweeplessSigFailed,
+				fsm.OnError:            SucceededSweeplessSigFailed,
+			},
+			Action: f.FetchSignPushSweeplessSweepTxAction,
+		},
+		HtlcTimeoutSwept: fsm.State{
+			Action: fsm.NoOpAction,
+		},
+		Succeeded: fsm.State{
+			Action: fsm.NoOpAction,
+		},
+		SucceededSweeplessSigFailed: fsm.State{
+			Action: fsm.NoOpAction,
+		},
+		UnlockDeposits: fsm.State{
+			Transitions: fsm.Transitions{
+				OnRecover:   UnlockDeposits,
+				fsm.OnError: Failed,
+			},
+			Action: f.UnlockDepositsAction,
+		},
+		Failed: fsm.State{
+			Action: fsm.NoOpAction,
+		},
+	}
+}
+
+// updateLoopIn is called after every action and updates the loop-in in the db.
+func (f *FSM) updateLoopIn(notification fsm.Notification) {
+	f.Infof("Current: %v", notification.NextState)
+
+	// Skip the update if the loop-in is not yet initialized.
+	if f.loopIn == nil {
```

Comment:
nit: I'd also add a comment here that this can happen as we call update on action entry.